Extension/src/main.ts

```diff
@@ -11,330 +11,135 @@ import * as util from './common';
-                        })
-                        .then(() => makeOfflineBinariesExecutable(installBlob))
-                        .then(() => rewriteManifest(installBlob))
-                        .then(() => touchInstallLockFile(installBlob))
-                        .catch(error => handleError(installBlob, error))
-                        .then(() => postInstall(installBlob))
-                        .then(() => activateExtensions());
-                } else {
-                    util.touchPackageLockFile();
-                    return activateExtensions();
-                }
-            } else {
-                // Need to watch for debugger.reload in case launch debugging is done.
-                fs.watch(extensionContext.extensionPath, (event: string, filename: string) => {
-                    if (filename == "debugger.reload") {
-                        util.checkInstallLockFile().then((installLockExists) => {
-                            if (installLockExists) {
-                                util.showReloadPrompt();
-                            } else {
-                                util.setDebuggerReloadLater();
-                                util.showWaitForDownloadPrompt();
-                            }
-                        });
-                    }
-                });
-                let channel: vscode.OutputChannel = util.getOutputChannel();
-                channel.appendLine("Updating C/C++ dependencies...");
-
-                let statusItem: vscode.StatusBarItem = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right);
-                let packageManager: PackageManager;
-
-                PlatformInformation.GetPlatformInformation()
-                    .then((info) => {
-                        installBlob.info = info;
-                        packageManager = new PackageManager(info, channel, statusItem);
-                        channel.appendLine("");
-                        installBlob.stage = "downloadPackages";
-                        return packageManager.DownloadPackages();
-                    })
-                    .then(() => {
-                        channel.appendLine("");
-                        installBlob.stage = "installPackages";
-                        return packageManager.InstallPackages();
-                    })
-                    .then(() => makeBinariesExecutable(installBlob))
-                    .then(() => removeUnnecessaryFile(installBlob))
-                    .then(() => rewriteManifest(installBlob))
-                    .then(() => touchInstallLockFile(installBlob))
-                    .catch(error => handleError(installBlob, error))
-                    .then(() => statusItem.dispose())
-                    .then(() => postInstall(installBlob))
-                    .then(() => activateExtensions());
-            }
-        });
-    });
+    let makeBinariesExecutablePromise: Promise<void> = makeBinariesExecutable();
```

Comment:
allowing the tasks to execute in arbitrary order breaks the usefulness of the installation "stage" telemetry.  The concept of adding an installation stage to the telemetry requires a deterministic execution order so we can know what has succeeded prior to the extension getting stuck.  (this applies to both online and offline installations)